Mu2eG4Filters/src/GammaConversionPoints_module.cc

```diff
@@ -0,0 +1,275 @@
+//
+// Largely based on Filters/src/InflightPionHits and Analyses/src/StoppedParticleFinder
+// Finds and writes out information for generated photons that pair produce
+// Writes conversion point x, y, z, t, px, py, pz, material (z, zeff, and fraction) , event weight, and gen energy
+
+// Mu2e includes.
+#include "MCDataProducts/inc/GenParticleCollection.hh"
+#include "MCDataProducts/inc/StatusG4.hh"
+#include "MCDataProducts/inc/StepPointMCCollection.hh"
+#include "MCDataProducts/inc/SimParticleCollection.hh"
+#include "MCDataProducts/inc/SimParticle.hh"
+#include "MCDataProducts/inc/SimParticleTimeMap.hh"
+#include "MCDataProducts/inc/EventWeight.hh"
+#include "MCDataProducts/inc/PhysicalVolumeInfoMultiCollection.hh"
+#include "Mu2eUtilities/inc/PhysicalVolumeMultiHelper.hh"
+#include "Mu2eG4/inc/findMaterialOrThrow.hh"
+#include "GeneralUtilities/inc/RSNTIO.hh"
+
+// G4 includes.
+#include "G4Material.hh"
+#include "G4Element.hh"
+#include "G4ElementVector.hh"
+
+// Framework includes.
+#include "canvas/Persistency/Common/Ptr.h"
+#include "art/Framework/Core/EDFilter.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art_root_io/TFileService.h"
+
+// Root includes
+#include "TH1F.h"
+#include "TH1D.h"
```

Comment:
On getting rid of TH1-s:  an example of using a text file that defines a tabulated spectrum is can be found here
https://github.com/Mu2e/Offline/blob/54cf3799c7dcc92a462303bb425ab60f460c88f3/EventGenerator/src/StoppedParticleReactionGun_module.cc#L121

Reply:
I'm not using TH1's for the spectra, but just for debug histogramming. I can remove these histograms if needed,
but they're useful in understanding the generated spectra as well as the event weights.

I'm using TF1's for the spectra. These are using arbitrary input spectrum formulas. Is the suggestion to not allow this type
of spectrum, or that I should generate a tabulated spectrum for any spectrum shape I choose to use?

From above, describing how I use the TF1's:
> Currently the TF1s are used to:
>
> - Compile the input spectrum strings (e.g. "max(100*(0.1-abs(x)), 0.1)" for cos(theta) and "1/((90-x)^4 + 0.1)" for energy)
> - Check that the minimum function value in the domain is >= 0 (PDF requirement)
> - Integrate the function over the relevant domain (normalize the PDF for event weights)
> - Find the maximum value in the domain (for accept/reject sampling)
> - Evaluate the spectrum at various points for sampling

Best,
Michael

Reply:
Hi Michael,

I misunderstood the statement in the last week's meeting.  I thought you  
were using a TH1 to represent an input spectrum from which to sample.
It looks like there are pieces of Mu2e code that use TF1.  I still find it ill-advised.
(May be from my experience with ROOT's integral of sin(x) from 0 to pi being zero
in a ROOT version,  then fixed, then becoming zero again later.)   Once you
are settled with the spectrum parameterization I think it's better to use C++ functions
rather than ROOT functions.   Your integrals are probably analytic.  See e.g.
https://github.com/Mu2e/Offline/blob/master/Mu2eUtilities/src/MuonCaptureSpectrum.cc